### PR TITLE
ethless: support state TxCosts

### PR DESCRIPTION
**Pull request by jordipainan**
THIS COMMIT BREAKS GENESIS on DEV and STAGE ENVIRONMENTS

### CODE REVIEW

**Review by NateWilliams2 — APPROVED** [review]
LGTM, I have one comment though-

Maybe this could be added in a different PR, but could we include `SetTreasurer`, `MintTokens`, and `SetTxCost` functions exposed in Vocone.go? This would be really helpful in testing, i.e. in the vaas-api I can use these to allocate tokens to a faucet account.

I implemented these methods locally to test with:
```
// SetTreasurer configures the vocone treasurer account address
func (vc *Vocone) SetTreasurer(treasurer common.Address) error {
	if err := vc.app.State.SetTreasurer(treasurer); err != nil {
		return err
	}
	vc.app.Commit()
	return nil
}

// MintTokens mints tokens to the given account address
func (vc *Vocone) MintTokens(to common.Address, amount uint64) error {
	if err := vc.app.State.MintBalance(to, amount); err != nil {
		return err
	}
	vc.app.Commit()
	return nil
}

// SetTxCost configures the transaction cost for the given tx type
func (vc *Vocone) SetTxCost(txType models.TxType, cost uint64) error {
	if err := vc.app.State.SetTxCost(txType, cost); err != nil {
		return err
	}
	vc.app.Commit()
	return nil
}
```

I included vc.App.Commit() so that this code can be run synchronously without waiting for the next block to be mined, etc. I understand if this is improper and shouldn't be used though. @jordipainan  and maybe @p4u what do u think?

**Comment by p4u** [comment]
> I included vc.App.Commit() so that this code can be run synchronously without waiting for the next block to be mined, etc. I understand if this is improper and shouldn't be used though. @jordipainan and maybe @p4u what do u think?

Yes, we can export that methods, I think that makes sense.

Regarding the app.Commit() no way :/ haha

**Comment by p4u** [comment]
> > I included vc.App.Commit() so that this code can be run synchronously without waiting for the next block to be mined, etc. I understand if this is improper and shouldn't be used though. @jordipainan and maybe @p4u what do u think?
> 
> Yes, we can export that methods, I think that makes sense.
> 
> Regarding the app.Commit() no way :/ haha

Can you send a suggestion so JP can just apply it?

File: cmd/vochaintest/vochaintest.go [hunk 1]
```diff
@@ -146,6 +147,11 @@ func main() {
 		censusImport(*host, entityKey)
 	case "censusGenerate":
 		censusGenerate(*host, entityKey, *electionSize, *keysfile, 1)
+	case "testtxs":
```
Comment (p4u) [inline cmd/vochaintest/vochaintest.go]: I think we can improve this naming to at least "testTxs", but I'd suggest "testTokenTransactions"

testAllTransactions should take the treasurer key not the oracle (even if its the same, for consistency the function signature should have the correct naming for input variables).

File: cmd/vochaintest/vochaintest.go [hunk 2]
```diff
@@ -947,3 +953,104 @@ func cspVoteTest(
 	}
 	log.Infof("all done!")
 }
+
+// enduser voting is not tested here
+func testAllTransactions(
+	host,
+	oraclePrivKey string,
```
Comment (p4u) [inline cmd/vochaintest/vochaintest.go]: that should be treasurerKey

File: cmd/vochaintest/vochaintest.go [hunk 3]
```diff
@@ -947,3 +953,104 @@ func cspVoteTest(
...
+) {
+	// oracle is also the treasurer
```
Comment (p4u) [inline cmd/vochaintest/vochaintest.go]: Why this assumption? You really don't know who will call this function

File: cmd/vochaintest/vochaintest.go [hunk 4]
```diff
@@ -947,3 +953,104 @@ func cspVoteTest(
...
+	oracleSigner.VocdoniChainID = chainId
+
+	// check set transaction cost
+	if err := testSetTxCost(mainClient, oracleSigner); err != nil {
+		log.Fatal(err)
+	}
+
+}
+
+func testSetTxCost(mainClient *client.Client, oracleSigner *ethereum.SignKeys) error {
```
Comment (p4u) [inline cmd/vochaintest/vochaintest.go]: oracleSigner should be treasurerSigner maybe?

**Comment by NateWilliams2** [comment]
> Regarding the app.Commit() no way :/ haha

Thought so but I had to try 😛

File: dockerfiles/testsuite/start_test.sh [hunk 5]
```diff
@@ -67,10 +74,18 @@ testid="/tmp/.vochaintest$RANDOM"
...
+[ $TEST -eq 5 -o $TEST -eq 0 ] && {
+	echo "### Running test 5 ###"
+	test_txs ${testid}5 &
+} || echo 0 >${testid}5
+
+echo "### Waiting for all state txs tests ###"
 wait
 
-[ "$(cat ${testid}1)" == "0" -a "$(cat ${testid}2)" == "0" -a "$(cat ${testid}3)" == "0" -a "$(cat ${testid}4)" == "0" ] && {
+[ "$(cat ${testid}1)" == "0" -a "$(cat ${testid}2)" == "0" -a "$(cat ${testid}3)" == "0" -a "$(cat ${testid}4)" == "0" -a "$(cat ${testid}5)" == "0" ] && {
```
Comment (p4u) [inline dockerfiles/testsuite/start_test.sh]: I think its time to break this line

File: rpcapi/accounthandlers.go [hunk 6]
```diff
@@ -31,10 +33,44 @@ func (r *RPCAPI) getAccount(request *api.APIrequest) (*api.APIresponse, error) {
 	response.Nonce = new(uint32)
```
Comment (p4u) [inline rpcapi/accounthandlers.go:32]: Regarding the previous lines (not part of this PR but related):

```
	// account does not exist
	if acc == nil {
		// response ok meaning the account does not exist
		return &response, nil
	}
```
I think we should return an error if the account does not exists.

File: rpcapi/accounthandlers.go [hunk 7]
```diff
@@ -31,10 +33,44 @@ func (r *RPCAPI) getAccount(request *api.APIrequest) (*api.APIresponse, error) {
...
 	*response.Nonce = acc.Nonce
 	response.InfoURI = acc.InfoURI
-	delegates := make([]string, len(acc.DelegateAddrs))
-	for i := 0; i < len(acc.DelegateAddrs); i++ {
-		delegates = append(delegates, common.BytesToAddress(acc.DelegateAddrs[i]).String())
+	if len(acc.DelegateAddrs) > 0 {
+		for _, v := range acc.DelegateAddrs {
+			response.Delegates = append(response.Delegates, common.BytesToAddress(v).String())
```
Comment (p4u) [inline rpcapi/accounthandlers.go:36]: Is not the same as doing `fmt.Sprintf("0x%x", v)` ?

File: rpcapi/accounthandlers.go [hunk 8]
```diff
@@ -31,10 +33,44 @@ func (r *RPCAPI) getAccount(request *api.APIrequest) (*api.APIresponse, error) {
...
+func (r *RPCAPI) getTreasurer(request *api.APIrequest) (*api.APIresponse, error) {
+	// try get treasurer and send reply
+	var response api.APIresponse
+	t, err := r.vocapp.State.Treasurer(true)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get treasurer: %w", err)
+	}
+	// treasurer does not exist
+	if t == nil {
+		// response ok meaning the treasurer does not exist
```
Comment (p4u) [inline rpcapi/accounthandlers.go]: Same, I think this should be an error

**Comment by NateWilliams2** [comment]
Interesting, It seems not possible to suggest changes to a file (vocone.go) that's not included in the original PR. I'll push a commit and ask Jordi to review it

File: rpcapi/accounthandlers.go [hunk 9]
```diff
@@ -31,10 +33,44 @@ func (r *RPCAPI) getAccount(request *api.APIrequest) (*api.APIresponse, error) {
...
+		return &response, nil
+	}
+	response.EntityID = common.BytesToAddress(t.Address).String()
```
Comment (p4u) [inline rpcapi/accounthandlers.go]: You can build response here

`response := api.APIresponse { EntityID : ... `

Also I think you don't need to call common.BytesToAddress

File: vocone/vocone.go [hunk 10]
```diff
@@ -195,6 +196,21 @@ func (vc *Vocone) AddOracle(oracleKey *ethereum.SignKeys) error {
 	return nil
 }
 
+// SetTreasurer configures the vocone treasurer account address
```
Comment (NateWilliams2) [inline vocone/vocone.go:199]: @jordipainan what do you think of these changes?

File: vochain/genesis.go [hunk 11]
```diff
@@ -288,12 +288,16 @@ var Genesis = map[string]VochainGenesis{
       ],
       "treasurer": "0x309Bd6959bf4289CDf9c7198cF9f4494e0244b7d",
       "tx_cost": {
-         "Tx_SetProcess": 100,
+         "Tx_SetProcessStatus": 100,
```
Comment (p4u) [inline vochain/genesis.go:291]: This is breaking for the Genesis.

If a commit modigy the Genesis it should update the genesis version and time and also add a comment to the commit message. Also I think stage genesis should be updated.

File: vochain/state.go [hunk 12]
```diff
@@ -504,39 +520,103 @@ func (v *State) IsOracle(addr common.Address) (bool, error) {
...
+// IsTreasurer returns true if the given address matches the Treasurer address
 func (v *State) IsTreasurer(addr common.Address) (bool, error) {
-	tAddr, err := v.Treasurer(false)
+	t, err := v.Treasurer(false)
 	if err != nil {
 		return false, err
 	}
-
+	tAddr := common.BytesToAddress(t.Address)
 	return addr == tAddr, nil
```
Comment (p4u) [inline vochain/state.go]: `return addr == common.BytesToAddress(t.Address), nil` or `bytes.Equal(addr, t.Address)`

File: vochain/state.go [hunk 13]
```diff
@@ -504,39 +520,103 @@ func (v *State) IsOracle(addr common.Address) (bool, error) {
...
 }
 
+// incrementTreasurerNonce increments the treasurer nonce
+func (v *State) incrementTreasurerNonce() error {
+	t, err := v.Treasurer(false)
+	if err != nil {
+		return err
```
Comment (p4u) [inline vochain/state.go]: Ideally all these errors should be wrapped. Else its very hard to identify where it comes from

File: vochain/state.go [hunk 14]
```diff
@@ -504,39 +520,103 @@ func (v *State) IsOracle(addr common.Address) (bool, error) {
...
+// SetTxCost sets the given transaction cost
+func (v *State) SetTxCost(txType models.TxType, cost uint64) error {
+	key, ok := TxTypeCostToStateKey[txType]
+	if !ok {
+		return fmt.Errorf("txType %v shouldn't cost anything", txType)
+	}
+	v.Tx.Lock()
+	defer v.Tx.Unlock()
+
+	costBytes := [8]byte{}
```
Comment (p4u) [inline vochain/state.go]: make is more idiomatic here

File: vochain/transaction.go [hunk 15]
```diff
@@ -585,3 +597,43 @@ func checkRevealProcessKeys(tx *models.AdminTx, process *models.Process) error {
 	}
 	return nil
 }
+
+func SetTransactionCostsTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (uint64, error) {
+	tx := vtx.GetSetTransactionCosts()
```
Comment (p4u) [inline vochain/transaction.go:618]: vtx should be checked before accessing fields since its a pointer

File: vochain/transaction.go [hunk 16]
```diff
@@ -585,3 +597,43 @@ func checkRevealProcessKeys(tx *models.AdminTx, process *models.Process) error {
...
+	// check signature available
+	if signature == nil || tx == nil || txBytes == nil {
+		return 0, fmt.Errorf("missing signature and/or transaction")
+	}
+	// check value
+	if tx.Value <= 0 {
```
Comment (p4u) [inline vochain/transaction.go]: This check removes the possibility to declare a tx with cost 0, I think we should allow that

File: vochain/transaction.go [hunk 17]
```diff
@@ -585,3 +597,43 @@ func checkRevealProcessKeys(tx *models.AdminTx, process *models.Process) error {
...
+		return 0, fmt.Errorf("cannot extract address from public key: %w", err)
+	}
+	// get treasurer
+	treasurer, err := state.Treasurer(false)
+	if err != nil {
+		return 0, err
+	}
+	// check signature recovered address
+	tAddr := common.BytesToAddress(treasurer.Address)
+	if tAddr != sigAddress {
```
Comment (p4u) [inline vochain/transaction.go]: nit: no need to declare tAddr

File: vochain/transaction.go [hunk 18]
```diff
@@ -585,3 +597,43 @@ func checkRevealProcessKeys(tx *models.AdminTx, process *models.Process) error {
...
+		return 0, fmt.Errorf("address recovered not treasurer: expected %s got %s", treasurer.String(), sigAddress.String())
+	}
+	// check nonce
+	if tx.Nonce != treasurer.Nonce {
```
Comment (p4u) [inline vochain/transaction.go:629]: I think these two last checks should go before extracting the pubkey from the signature (which is the expensive check)

File: vochain/transaction.go [hunk 19]
```diff
@@ -585,3 +597,43 @@ func checkRevealProcessKeys(tx *models.AdminTx, process *models.Process) error {
 	}
 	return nil
 }
+
+func SetTransactionCostsTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (uint64, error) {
```
Comment (p4u) [inline vochain/transaction.go:614]: Please document all exported methods :)

File: vochain/types.go [hunk 20]
```diff
@@ -59,13 +60,87 @@ type QueryData struct {
...
+	"SendTokens":              models.TxType_SEND_TOKENS,
+	"SetAccountInfo":          models.TxType_SET_ACCOUNT_INFO,
+	"RegisterKey":             models.TxType_REGISTER_VOTER_KEY,
+	"NewProcess":              models.TxType_NEW_PROCESS,
+	"AddDelegateForAccount":   models.TxType_ADD_DELEGATE_FOR_ACCOUNT,
+	"DelDelegateForAccount":   models.TxType_DEL_DELEGATE_FOR_ACCOUNT,
+	"CollectFaucet":           models.TxType_COLLECT_FAUCET,
+}
+
+func TxCostNameToTxType(key string) models.TxType {
```
Comment (p4u) [inline vochain/types.go:112]: document me!

**Review by p4u — CHANGES_REQUESTED** [review]

**Comment by jordipainan** [comment]
@p4u I cannot comment on your suggestions, regarding the one left (`Is not the same as doing fmt.Sprintf("0x%x", v) ?`) is not the same cause this does not convert to the checksum addr

**Comment by jordipainan** [comment]
seems the last integration test is not executed, do not merge until solved

**Review by p4u — APPROVED** [review]

**Comment by p4u** [comment]
great job

**Comment by NateWilliams2** [comment]
@p4u @jordipainan I realized a problem with the `SetTreasurer`, `MintTokens`, and `SetTxCost` functions I proposed. These methods all rely on making some changes to the state, so if they are called directly in the code from Vocone.go (without a hacky solution like manually calling `app.State.Save()`), they will have no effect. The next time a block is mined, the state will roll back, and then transactions from the mempool will be applied. The effects of calling these methods from `Vocone.go` will be discarded altogether. It could still be helpful to have these methods in the vocone, but maybe they can add an actual tx to the mempool rather than calling the methods directly. 

I'm actually confused because there are tests in vocdoni-node that seem to rely on calling methods like `app.State.SetTxCost` directly, without committing the state.

**Comment by jordipainan** [comment]
@NateWilliams2  
You're right, I thought you are going to call Save() later.
I also thought that the vocone is just for testing, so you can call Save() later or do not call it if you do not need it. We can talk more about this because maybe I miss some information.

The testing thing you mention is done for setting up the testing easily and does not impact on the actual test itself.

**Comment by NateWilliams2** [comment]
> I thought you are going to call Save() later

In this case, we need to expose `app.State.Save` to the vocone as well

File: vocone/vocone.go [hunk 21]
```diff
@@ -195,6 +196,40 @@ func (vc *Vocone) AddOracle(oracleKey *ethereum.SignKeys) error {
...
+		log.Debugf("setting tx cost for txtype %s", models.TxType_name[int32(k)])
+		err := vc.app.State.SetTxCost(k, txCosts)
+		if err != nil {
+			return err
+		}
+
+	}
+	return vc.app.State.IncrementTreasurerNonce()
+}
+
```
Comment (NateWilliams2) [inline vocone/vocone.go:254]:
```suggestion
func (vc *Vocone) SaveState() error {
	_, err := vc.app.State.Save()
	return err
}
```

File: vochain/state.go [hunk 22]
```diff
@@ -504,37 +520,100 @@ func (v *State) IsOracle(addr common.Address) (bool, error) {
...
+	t, err := v.Treasurer(false)
 	if err != nil {
 		return false, err
 	}
+	return addr == common.BytesToAddress(t.Address), nil
+}
 
-	return addr == tAddr, nil
+// IncrementTreasurerNonce increments the treasurer nonce
+func (v *State) IncrementTreasurerNonce() error {
```
Comment (p4u) [inline vochain/state.go:569]: does this method needs to be exported?

**Review by p4u — APPROVED** [review]

**Comment by randomshinichi** [comment]
> I'm actually confused because there are tests in vocdoni-node that seem to rely on calling methods like app.State.SetTxCost directly, without committing the state.

In my original functions, I was calling Save(), I just thought that since the functions you added here didn't have them, you had some solution of your own.

**Comment by jordipainan** [comment]
@p4u can I merge it ? I think we can squash everything in one commit

**Comment by p4u** [comment]
> @p4u can I merge it ? I think we can squash everything in one commit

Go ahead, ok to squash

**Pull request merged**